### ISSUE TITLE
Nodes are ready even if GCE routes are not created for them.

### ISSUE THREAD

**Issue by gmarek** [kubernetes/kubernetes#23327]
Observed in v1.2 release. Node is Ready even if route controller fails to create a route for a Node, which allows scheduler to schedule new Pods on it. Those Pods will be running, but the networking for them won't work, which makes it a bad experience for users. Node should be NotReady before route is created for it.

cc @kubernetes/goog-control-plane @kubernetes/sig-node @cjcullen @alex-mohr @roberthbailey

**Comment by gmarek** [comment]
@cjcullen - assigning to you, as I believe you're the one who's most familiar with RouteController logic.

**Comment by gmarek** [comment]
cc @davidopp @alex-mohr - do we treat this as a blocker for 1.3 as well? This has the same effect as a crashlooping kube-proxy which we treat as one.

**Comment by alex-mohr** [comment]
Seems reasonable to me as a P1 bug.

**Comment by zmerlynn** [comment]
This is actually much worse on larger clusters, though I think we need some work in the route controller to fix "Rate Limit Exceeded" errors there as well. We eventually reconcile, but we can do better.

**Comment by gmarek** [comment]
I don't have any idea for an easy fix - NodeProblem API would be perfect, but I don't think it's mature enough and is supported by the scheduler @dchen1107 @davidopp.

If we decide to fix it for 1.3 and the only way of making the node unschedulable is to set its state to NotReady (please tell me if that's no longer the case), then we'd need to have RouteController to update NodeStatus and make sure that Kubelet and NodeController won't update it.

Another approach would be to create a "reflector Pod" (not to mistake with mirror pods...) and Kubelet would need to receive some data from it,thus confirming that the networking is working, before it's allowed to send first "NodeReady" update. It shouldn't be a huge change, but I don't know the Kubelet code well enough. @dchen1107 @yujuhong

**Comment by cjcullen** [comment]
The Ready/NotReady logic in the NodeController is pretty complex. I'd hesitate to let the RouteController also modify the Ready condition. Would there be a way to let the RouteController set something in the NodeStatus that the NodeController would be waiting for?

I will not have time to own this for 1.3. Any volunteers?

**Comment by wojtek-t** [comment]
I agree with @cjcullen that letting RouteController modify Ready condition is risky.

>
Would there be a way to let the RouteController set something in the NodeStatus that the NodeController would be waiting for?

That's an interesting question. Can we introduce another condition (like NetworkReady or sth like that) and that can be set only by RouteController?
The NodeController may potentially propagate this information to Ready condition.

This change should be relatively simple to do, but I'm not sure if we want such API.
Thoughts?

**Comment by dchen1107** [comment]
Today's NodeProblemDetetor only detects kernel related issues, not other issues. We could extend it to report docker related problem for 1.3 release timeline. Detecting software bugs shouldn't be NodeProblemDetector's goal, especially the pieces / components controlled by us. Please not abuse it. 

Also please not create a "reflector Pod" in fly to complicate the logic since RouteController already has route creation status related to a given node. Why not follow what we have done through NodeProblemAPI by introduce a new Condition to NodeStatus, called RouteReady or something similar, and RouteController maintain that Condition? 

cc/ @bgrant0607 @derekwaynecarr  This is related to our recent NodeCondition discussion? If we should treat each condition as a problem?

**Comment by zmerlynn** [comment]
@davidopp asked separately about priority (offline), answering here. The priority of this bug is inversely related to #26119. If we can get route programming on large clusters down to O(minutes), this bug is back to being a nuisance that we can grit our teeth and ship without. As it is, after a little work on #26119, we still have laggard routes that aren't getting created for a *while*.

**Comment by wojtek-t** [comment]
@dchen1107 - what you suggested is exactly what I suggested in the post above yours. So it seems that at least we have a similar view on it.

**Comment by gmarek** [comment]
NC does not set 'ready = true' status, Kubelet does. NC sets only false and unknown statuses. I think that adding the new Status object and making NC/Kubelet pipe it to the NodeStatus is similar to setting NodeStatus in RC in terms of the risk.

As for reflector - we'd need to do some kind of network-based health at some point. Doing it in RC makes it GCE/GKE specific, which is fine, as it's easier. OTOH writing a reflector would solve this problem for all possible network configurations.

All things considered I think that adding new "NetworkConfigured" condition seems fine. This might get piped as a NodeReady status, or we can have scheduler skip "NetworkConfigured = false" Nodes when scheduling.

**Comment by gmarek** [comment]
We probably should think about renaming "NodeReady" to "NodeReachable" or sth if we want to add more Conditions.

**Comment by wojtek-t** [comment]
>
This might get piped as a NodeReady status, or we can have scheduler skip "NetworkConfigured = false" Nodes when scheduling.

Yeah - this is more a technical detail at this point.

>
We probably should think about renaming "NodeReady" to "NodeReachable" or sth if we want to add more Conditions.

I don't understand this one. At the level of API you have "NodeCondition". Currently one condition is "NodeReady". What we are suggesting is add additional "NodeReachable" (or whatever we call it). I don't understand what do you want to rename.

**Comment by gmarek** [comment]
If we want to add a new condition for route rediness (however we call it), then in practice "NodeReady" understood as "Node is ready to run Pods" would be a conjunction of current "NodeReady=true" (which really means that the Node is reachable now) and "RouteReady=true". I think that may be confusing.

**Comment by bgrant0607** [comment]
Schedulers shouldn't need to reason about any new conditions directly.

**Comment by bgrant0607** [comment]
And changing Ready would be a non-backward-compatible API change, as well.

**Comment by bgrant0607** [comment]
Conditions have reasons, so NoPodNetworking or somesuch could be a reason for Ready=false.

**Comment by gmarek** [comment]
@bgrant0607 - IIUC you're saying that we should add a new Condition and something should pipe it to NodeReady?

**Comment by zmerlynn** [comment]
Wait, things might have changed...
```
I0525 05:40:29.842539    3399 kubelet.go:2501] skipping pod synchronization - [Kubenet does not have netConfig. This is most likely due to lack of PodCIDR]
```

I just brought up a 1000 node cluster, and due to some other issue, I only have 136 routes and 136 `Ready` nodes. The rest have the above log message and no GCE route assigned. Are we doing the right thing?

**Comment by gmarek** [comment]
No, we're not. This is the situation when NC didn't assign CIDR to the Node. NC and RC do not communicate directly, but RC need PodCIDR to be assigned to create routes.

In other words - this failure is an earlier one, Node will become ready when PodCIDR will be allocated and route creation may still fail.

**Comment by zmerlynn** [comment]
Oh, my cluster is just hosed for a different reason. That's good. :)

**Comment by bgrant0607** [comment]
@gmarek I take back what I wrote before. I'll comment further on the PR.

**Comment by davidopp** [comment]
The PR in question is #26267 (it's not referenced anywhere else here).

**Comment by gmarek** [comment]
Fixed by #26415
